ui/app/hooks/useRetryTransaction.js

```diff
@@ -0,0 +1,110 @@
+import { useDispatch } from 'react-redux'
+import { useMemo, useCallback, useEffect, useState } from 'react'
+import * as actions from '../store/actions'
+import {
+  fetchBasicGasAndTimeEstimates as fetchBasicGasAndTimeEstimatesAction,
+  fetchGasEstimates as fetchGasEstimatesAction,
+  setCustomGasPriceForRetry as setCustomGasPriceForRetryAction,
+  setCustomGasLimit as setCustomGasLimitAction,
+} from '../ducks/gas/gas.duck'
+import { TOKEN_METHOD_TRANSFER } from '../helpers/constants/transactions'
+import { increaseLastGasPrice } from '../helpers/utils/confirm-tx.util'
+import { useMetricEvent } from './useMetricEvent'
+import { useMethodData } from './useMethodData'
+
+
+/**
+ * useRetryTransaction
+ *
+ * Provides a reusable hook that, given a transactionGroup, will return
+ * whether or not the transaction meets the criteria to be retried with
+ * higher gas (sped up), and a method for beginning the retry process
+ * @param {Object} transactionGroup - the transaction group
+ * @param {boolean} isEarliestNonce - indicates if this transaction is the earliest in history
+ * @return {[boolean, Function]}
+ */
+export function useRetryTransaction (transactionGroup, isEarliestNonce = false) {
+  const { primaryTransaction, transactions, initialTransaction, hasRetried } = transactionGroup
+  const [earliestTransaction = {}] = transactions
+  const { submittedTime } = earliestTransaction
+  const gasPrice = primaryTransaction.txParams?.gasPrice
+  const methodData = useMethodData(primaryTransaction.txParams?.data)
+  const trackMetricsEvent = useMetricEvent(({
+    eventOpts: {
+      category: 'Navigation',
+      action: 'Activity Log',
+      name: 'Clicked "Speed Up"',
+    },
+  }))
+  const dispatch = useDispatch()
+  const { fetchBasicGasAndTimeEstimates, fetchGasEstimates, setSelectedToken, retryTransaction } = useMemo(() => ({
+    fetchBasicGasAndTimeEstimates: () => dispatch(fetchBasicGasAndTimeEstimatesAction()),
+    fetchGasEstimates: (blockTime) => dispatch(fetchGasEstimatesAction(blockTime)),
+    setSelectedToken: (tokenAddress) => dispatch(actions.setSelectedToken(tokenAddress)),
+    retryTransaction: () => {
+      const transaction = initialTransaction
+      const increasedGasPrice = increaseLastGasPrice(gasPrice)
+      dispatch(setCustomGasPriceForRetryAction(increasedGasPrice || transaction.txParams?.gasPrice))
+      dispatch(setCustomGasLimitAction(transaction.txParams?.gas))
+      dispatch(actions.showSidebar({
+        transitionName: 'sidebar-left',
+        type: 'customize-gas',
+        props: { transaction },
+      }))
+    },
+  }), [dispatch, initialTransaction, gasPrice])
+  const [retryEnabled, setRetryEnabled] = useState(() => {
+    return Date.now() - submittedTime > 5000 && isEarliestNonce && !hasRetried
+  })
+
+  if (methodData?.name === TOKEN_METHOD_TRANSFER && initialTransaction.txParams?.to) {
```

Comment:
I think you're right it should be moved into the callback. Though I'm not entirely sure what the original purpose of this call was here: https://github.com/MetaMask/metamask-extension/blob/fd170ef8c0404b5260a36dba38245e629949c6d5/ui/app/components/app/transaction-list-item/transaction-list-item.component.js#L111-L125

Reply:
why would you set a selected token as a result of issuing a retry?

Reply:
That is an excellent question 😅 😬 

I'm not sure. This is one of the things I had planned to remove actually - in my asset page PR (I need to replace all uses of `selectedToken` with a token prop that gets passed in).

It's possible we're using the selected token somewhere in the retry logic? That doesn't make sense though 🤔. I'll look into this... maybe this can be omitted now...

Reply:
@Gudahtt i didn't touch this on this round of fixes.. lemme know if we're good to either remove or move it into callback